### PR TITLE
add: concurrent download

### LINKED ISSUE #7: 多线程下载
- [`Promise.all()`](https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Promise/all)

### PR DESCRIPTION
**Pull request by PapayaHUANG**
closes #7

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -27,6 +27,20 @@ async function getSubPageURLs(filePath: string, selector: string) {
...
+      await downloadFn(url, true);
+    });
+    await Promise.all(promises);
+  };
+
+  for (let i = 0; i < urls.length; i += concurrencyLimit) {
+    const batch = urls.slice(i, i + concurrencyLimit);
+    await downloadBatch(batch);
+  }
+};
```
Comment (wangrunlin) [inline src/index.ts:42]:
```suggestion
const downloadWithConcurrency = async (urls, concurrencyLimit = 5) => {
  for (let i = 0; i < urls.length; i += concurrencyLimit) {
    await Promise.all(urls.slice(i, i + concurrencyLimit).map(async (url) => {
      await downloadHTML(url, true);
    }));
  }
};
```

File: src/index.ts [hunk 2]
```diff
@@ -36,15 +50,15 @@ const root = new URL(url).pathname.slice(1);
...
 
   if (useless) await $`mv ${oldName} ${newName}`;
 
   if (!/\.html?$/.test(name)) continue;
 
-  const linkList = await getSubPageURLs(newName, "tr > td:nth-child(2) > a");
+  const linkList = await getSubPageURLs(newName, 'tr > td:nth-child(2) > a');
 
-  for (const link of linkList) await downloadHTML(link, true);
+  await downloadWithConcurrency(linkList, 5, downloadHTML);
```
Comment (wangrunlin) [inline src/index.ts:63]:
```suggestion
  await downloadWithConcurrency(linkList);
```

**Review by wangrunlin — CHANGES_REQUESTED** [review]

**Review by wangrunlin — APPROVED** [review]

**Comment by wangrunlin** [comment]
还需要解决冲突

File: src/index.ts [hunk 3]
```diff
@@ -27,6 +27,16 @@ async function getSubPageURLs(filePath: string, selector: string) {
...
 }
 
+const downloadWithConcurrency = async (urls, concurrencyLimit = 5) => {
+  for (let i = 0; i < urls.length; i += concurrencyLimit) {
+    await Promise.all(
+      urls.slice(i, i + concurrencyLimit).map(async (url) => {
+        await downloadHTML(url, true);
+      })
+    );
+  }
```
Comment (TechQuery) [inline src/index.ts:37]:
```suggestion
const downloadWithConcurrency = async (urls: string[], concurrencyLimit = 5) => {
  for (const list of splitArray(urls, concurrencyLimit))
    await Promise.all(
      list.map((url) => downloadHTML(url, true))
    ):
```
https://web-cell.dev/web-utility/functions/splitArray.html

File: src/index.ts [hunk 4]
```diff
@@ -1,12 +1,12 @@
 #! /usr/bin/env node
 
-import { JSDOM } from "jsdom";
-import { $, argv, cd, fs, path } from "zx";
+import { JSDOM } from 'jsdom';
+import { $, argv, cd, fs, path } from 'zx';
 
 const [url, folder] = argv._;
```
Comment (TechQuery) [inline src/index.ts:6]: 可选的并发参数也要传入，上游库：
https://github.com/TechQuery/CommanderJSX

**Review by TechQuery — CHANGES_REQUESTED** [review]

**Comment by PapayaHUANG** [comment]
> 还需要解决冲突

好滴～

File: src/index.tsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+Command.execute(
+  <Command
+    parameters="[options] <url> <folder>"
+    options={{
+      concurrencyLimit: {
+        shortcut: 'c',
+        parameters: '<number>',
+        pattern: /\d+/,
```
Comment (TechQuery) [inline src/index.tsx:88]:
```suggestion
        pattern: /^\d+$/,
```

File: tsconfig.json [hunk 6]
```diff
@@ -3,7 +3,9 @@
     "target": "ES2017",
     "module": "NodeNext",
     "moduleResolution": "NodeNext",
+    "jsx": "react",
+    "jsxFactory": "createCommand",
     "outDir": "dist/"
   },
-  "include": ["src/**/*.ts"]
+  "include": ["src/index.tsx", "src/**/*.tsx"]
```
Comment (TechQuery) [inline tsconfig.json:10]:
```suggestion
  "include": ["src/**/*"]
```

**Review by TechQuery — CHANGES_REQUESTED** [review]

**Review by TechQuery — APPROVED** [review]

**Pull request merged**